File: utilities/point.py
import numpy as np
import math


class Point:
    """
    Mỗi điểm trong không gian R^d có các thông tin:
    - axes: Vector tọa độ d chiều (dùng cấu trúc mảng của NumPy)
    - steiner: Biến nhị phân, cho biết đây là điểm Steiner hay terminal
    VD: Điểm terminal A(1.0, 2.0, 3.0) có:
    - axes = np.array([1.0, 2.0, 3.0])
    - steiner = False
    """

    def __init__(self, axes=None, steiner=False):
        self.axes = axes
        self.steiner = steiner

    def __hash__(self):
        return hash(id(self))

    def equals(self, p):
        return np.array_equal(self.axes, p.axes)

    @classmethod
    def from_line(cls, line: str):
        axes = [float(axis) for axis in line.split()]
        return cls(np.array(axes))

    def set_axes(self, axes):
        self.axes = axes


# Tính khoảng cách euclidean giữa 2 điểm
def distance(A, B):
    return np.linalg.norm(A.axes - B.axes)


# Tính tích vô hướng 2 vector BA và BC
def dot_product(A, B, C):
    return np.dot(A.axes - B.axes, C.axes - B.axes)


# Tính góc giữa 2 vector BA và BC theo độ đo radiant
def angle(A, B, C):
    if np.array_equal(A.axes, B.axes) or np.array_equal(A.axes, C.axes) or np.array_equal(C.axes, B.axes):
        return math.pi
    norm_BA = (B.axes - A.axes) / np.linalg.norm(B.axes - A.axes)
    norm_BC = (B.axes - C.axes) / np.linalg.norm(B.axes - C.axes)
    return np.arccos(np.clip(np.dot(norm_BA, norm_BC), -1, 1))


# Tính trọng tâm của 1 tập điểm
def center_of_gravity(points: list):
    point_axes = [p.axes for p in points]
    return np.sum(point_axes, axis=0) / len(points)


# Tính điểm Fermat của tam giác
def fermat_point(points):
    if len(points) != 3:
        return None

    A, B, C = points

    # Kiểm tra 1 trong 3 góc > 120 độ, trùng điểm Fermat
    angle_A = angle(B, A, C)
    if angle_A >= math.pi * 2 / 3:
        return A.axes
    angle_B = angle(A, B, C)
    if angle_B >= math.pi * 2 / 3:
        return B.axes
    angle_C = angle(A, C, B)
    if angle_C >= math.pi * 2 / 3:
        return C.axes

    # Tính điểm Fermat theo công thức vector
    a = distance(B, C)
    b = distance(A, C)
    c = distance(A, B)

    x = 1 / math.sin(angle_A + math.pi / 3)
    y = 1 / math.sin(angle_B + math.pi / 3)
    z = 1 / math.sin(angle_C + math.pi / 3)

    _D = a*x+b*y+c*z

    return (a*x/_D) * A.axes + (b*y/_D) * B.axes + (c*z/_D) * C.axes


# if __name__ == '__main__':
#     A = Point(np.array([1.0, 3.0, 1.0, 2.0]))
#     B = Point(np.array([2.0, 3.0, 2.0, 2.0]))
#     C = Point(np.array([1.0, 3.0, 2.0, 2.0]))
#     F = Point(fermat_point([A, B, C]))
#     print(F.axes)
#     ab = distance(A, B)
#     bc = distance(B, C)
#     ca = distance(C, A)
#
#     print(distance(F, A) + distance(F, B) + distance(F, C))
#     print(ab + bc)
#     print(bc + ca)
#     print(ab + ca)

File: cpp1/point.cpp
// point.cpp
#include "point.hpp"
#include <sstream>
#include <algorithm>
#include <numeric>

Point::Point(const std::vector<double>& axes, bool steiner) : axes(axes), steiner(steiner) {}

bool Point::equals(const Point& p) const {
    return axes == p.axes;
}

Point Point::from_line(const std::string& line) {
    std::istringstream iss(line);
    std::vector<double> axes;
    double axis;
    while (iss >> axis) {
        axes.push_back(axis);
    }
    return Point(axes);
}

void Point::set_axes(const std::vector<double>& axes) {
    this->axes = axes;
}

double distance(const Point& A, const Point& B) {
    double sum = 0.0;
    for (size_t i = 0; i < A.axes.size(); ++i) {
        sum += std::pow(A.axes[i] - B.axes[i], 2);
    }
    return std::sqrt(sum);
}

double dot_product(const Point& A, const Point& B, const Point& C) {
    double sum = 0.0;
    for (size_t i = 0; i < A.axes.size(); ++i) {
        sum += (A.axes[i] - B.axes[i]) * (C.axes[i] - B.axes[i]);
    }
    return sum;
}

double angle(const Point& A, const Point& B, const Point& C) {
    if (A.axes == B.axes || A.axes == C.axes || C.axes == B.axes) {
        return M_PI;
    }
    std::vector<double> BA(A.axes.size()), BC(A.axes.size());
    for (size_t i = 0; i < A.axes.size(); ++i) {
        BA[i] = B.axes[i] - A.axes[i];
        BC[i] = B.axes[i] - C.axes[i];
    }
    double norm_BA = std::sqrt(std::inner_product(BA.begin(), BA.end(), BA.begin(), 0.0));
    double norm_BC = std::sqrt(std::inner_product(BC.begin(), BC.end(), BC.begin(), 0.0));
    double dot = std::inner_product(BA.begin(), BA.end(), BC.begin(), 0.0) / (norm_BA * norm_BC);
    return std::acos(std::max(-1.0, std::min(1.0, dot)));
}

std::vector<double> center_of_gravity(const std::vector<Point>& points) {
    std::vector<double> result(points[0].axes.size(), 0.0);
    for (const auto& p : points) {
        for (size_t i = 0; i < p.axes.size(); ++i) {
            result[i] += p.axes[i];
        }
    }
    for (auto& r : result) {
        r /= points.size();
    }
    return result;
}

std::vector<double> fermat_point(const std::vector<Point>& points) {
    if (points.size() != 3) {
        return {};
    }

    const Point& A = points[0];
    const Point& B = points[1];
    const Point& C = points[2];

    double angle_A = angle(B, A, C);
    if (angle_A >= 2 * M_PI / 3) {
        return A.axes;
    }
    double angle_B = angle(A, B, C);
    if (angle_B >= 2 * M_PI / 3) {
        return B.axes;
    }
    double angle_C = angle(A, C, B);
    if (angle_C >= 2 * M_PI / 3) {
        return C.axes;
    }

    double a = distance(B, C);
    double b = distance(A, C);
    double c = distance(A, B);

    double x = 1 / std::sin(angle_A + M_PI / 3);
    double y = 1 / std::sin(angle_B + M_PI / 3);
    double z = 1 / std::sin(angle_C + M_PI / 3);

    double D = a * x + b * y + c * z;

    std::vector<double> result(A.axes.size());
    for (size_t i = 0; i < result.size(); ++i) {
        result[i] = (a * x * A.axes[i] + b * y * B.axes[i] + c * z * C.axes[i]) / D;
    }
    return result;
}